c10/core/impl/cow/simulator.cpp

```diff
@@ -0,0 +1,33 @@
+#include <c10/core/impl/cow/simulator.h>
+
+#include <c10/util/Exception.h>
+
+#include <limits>
+
+namespace c10::impl {
+
+template <bool intrusive>
+cow::SimulatorImpl<intrusive>::SimulatorImpl(
+    std::uint64_t storage_generation) noexcept
+    : storage_generation_(storage_generation) {}
+
+template <bool intrusive>
+auto cow::SimulatorImpl<intrusive>::storage_generation() noexcept
+    -> std::uint64_t {
+  std::lock_guard<std::mutex> lock(mtx_);
+  return storage_generation_;
+}
+
+template <bool intrusive>
+auto cow::SimulatorImpl<intrusive>::bump_storage_generation() noexcept
+    -> std::uint64_t {
+  std::lock_guard<std::mutex> lock(mtx_);
+  TORCH_INTERNAL_ASSERT(
+      storage_generation_ != std::numeric_limits<std::uint64_t>::max());
+  return ++storage_generation_;
```

Comment:
Done.

It should have always been atomic. Early in the design I was being cautious about ensuring that we always write and read both generation numbers atomically. I turned out to be overly cautious, since we update them both under the physical storage's mutex and we always have that mutex held whenever we need to read them atomically.